README draft (#64)
fixie-ai/ai-jsx: readme.md

Reviewer: If you wanted, you could add an analogy to database query construction. Although it's technically possibly to construct a db query via string templating, you'd be wild to do so. This is the same higher degree of support for prompts. (If you wanted to take it further, you could argue that LLMs are stochastic or "soft" databases and prompts are the queries for them.)

<p align="center">
  <img src="https://docs.ai-jsx.com/img/ai-jsx.png" alt="AI.JSX logo" width="100" />
</p>

# AI.JSX — The AI Application Framework for Javascript

[![Docs Site](https://img.shields.io/badge/Docs%20Site-docs.ai--jsx.com-orange)](https://docs.ai-jsx.com)
[![Discord Follow](https://dcbadge.vercel.app/api/server/MsKAeKF8kU?style=flat)](https://discord.gg/MsKAeKF8kU)

> [![Twitter Follow](https://img.shields.io/twitter/follow/fixieai?style=social)](https://twitter.com/fixieai)

Reviewer: https://youtu.be/cP-rQb_daYE?t=63

Reviewer: There are more of these, if you care to add them, for weekly npm installs, listing the latest version on npm, etc.

> AI.JSX is a framework for building AI applications using [JSX](https://react.dev/learn/writing-markup-with-jsx). With AI.JSX, you don't just use JSX to describe what your UI should look like, you also use it to describe how **AI (LLMS)** should integrate into the rest of your application. The end result is a powerful combination where _intelligence_ can be deeply embedded into the application stack.

Reviewer: ```suggestion
AI.JSX is a framework for building AI applications using [JSX](https://react.dev/learn/writing-markup-with-jsx). With AI.JSX, you don't just use JSX to describe what your UI should look like, you also use it to describe how **AI (LLMs)** should integrate into the rest of your application. The end result is a powerful combination where _intelligence_ can be deeply embedded into the application stack.
```

AI.JSX is designed to give you two important capabilities out of the box:

> 1. A powerful mechanism for orchestrating multiple LLM calls expressed as modular, re-usable components.

Reviewer: Nit: it would be better not to reuse the word "powerful"

> 1. The ability to seamlessly interweave UI components with your AI compoents. This means you can rely on the LLM to construct your UI dynamically from a set of standard React components.

Reviewer: ```suggestion
1. The ability to seamlessly interweave UI components with your AI compoents. This means you can rely on the LLM to construct your UI dynamically from a set of React components you provide.
```

"Standard" makes it sound like "there's a standard library", rather than "well-known components you provide".

AI.JSX can be used to create standalone LLM applications that can be deployed anywhere Node.JS is supported, or it can be used as part of a larger React application. We have examples that use [Next.js](https://github.com/fixie-ai/ai-jsx/tree/main/packages/nextjs-demo) and [Create React App](https://github.com/fixie-ai/ai-jsx/tree/main/packages/create-react-app-demo). For more details, see our [AI+UI guide](https://docs.ai-jsx.com/guides/ai-ui).

> To get started, see the quickstart below, read our [Getting Started Guide](https://docs.ai-jsx.com/getting-started), or check out the [examples](https://github.com/fixie-ai/ai-jsx/tree/main/packages/examples).

Reviewer: I would also add a link to https://docs.ai-jsx.com/guides/brand-new, since many JS devs will be totally new to all these concepts. ("What's an LLM? How could I possibly use it to drive application behavior?")

To learn more about why we built AI.JSX, you can read our [blog post](https://blog.fixie.ai).

## Quickstart

> The fastest way to get started is to [clone our template repo](https://github.com/fixie-ai/ai-jsx-template). But here's how to do it manually:

Reviewer: I would also add a link to the "hello world" code sandbox, either here or in the "to get started" paragraph above.

Create your folder and `cd` into it:

```console
mkdir ai-jsx-example && cd ai-jsx-example
```

Install Typescript:

```console
npm i typescript --save-dev
```

Initialize a new Typescript project:

```console
npx tsc --init
```

Install `ai-jsx`:

```console

> npm i ai-jsx --save-dev

Reviewer: This should not be a dev dep, since it will be used at runtime.

```

> Set up your `package.json`:

Reviewer: We should also advise people to set their `tsconfig.json` to match what's in the template repos.

  },
  "devDependencies": {
    "@tsconfig/node18": "^2.0.1",
    "tsx": "^3.12.7",
    "typescript": "^5.1.3"
  }
}
```

Create a new `index.tsx` file with the following:

```tsx
import * as LLMx from 'ai-jsx';
import { ChatCompletion, SystemMessage, UserMessage } from 'ai-jsx/core/completion';

function App() {
  return (
    <ChatCompletion>
      <SystemMessage>You are an agent that only gives snarky, sarcastic advice.</SystemMessage>
      <UserMessage>How can I learn about Ancient Egypt?</UserMessage>
    </ChatCompletion>
  );
}

console.log(await LLMx.createRenderContext().render(<App />));
```

## Features

- Prompt engineering through modular, reusable components
- The ability to easily switch between model providers and LLM configurations (e.g., temperature)
- Built in support for Tools (ReAct pattern), Document Question and Answering, Chain of Thought, and more
- Ability to directly interweave LLM calls with standard UI components, including the ability for the LLM to render the UI from a set of provided components
- Built-in streaming support
- First-class support for NextJS and Create React App (more coming soon)
- Full support for LangChainJS

## Examples

> For a full set of examples, see (here)[https://github.com/fixie-ai/ai-jsx/tree/main/packages/examples].

Reviewer: ![image](https://github.com/fixie-ai/ai-jsx/assets/829827/7f1fa850-b18c-4c51-bad1-247ddfcc05b1)

```suggestion
For a full set of examples, see [the examples package](https://github.com/fixie-ai/ai-jsx/tree/main/packages/examples).
```
